mountpoint-s3/src/inode.rs

```diff
@@ -540,12 +581,45 @@ impl ReaddirStreamState {
 
 impl ReaddirHandle {
     pub async fn next<OC: ObjectClient>(&self, client: &OC) -> Result<Option<LookedUp>, InodeError> {
-        while self.results.read().unwrap().is_empty() {
+        // We will start fetching new results when number of items in the queue is less than number of local results
+        // so that we can make sure the order is correct
+        while self.results.read().unwrap().len() <= self.local_results.read().unwrap().len() {
             let continuation_token = {
                 let mut next_token = self.next_continuation_token.lock().unwrap();
+
+                // append local files before the first stream
+                if *next_token == ReaddirStreamState::NotStarted {
+                    let inode = self.inner.get(self.dir_ino)?;
+                    let kind_data = &inode.inner.sync.read().unwrap().kind_data;
+                    let local_files = match kind_data {
+                        InodeKindData::File { .. } => unreachable!("we know this is a directory"),
+                        InodeKindData::Directory {
+                            children: _,
+                            writing_children,
+                        } => writing_children.iter().map(|ino| {
+                            let inode = self.inner.get(*ino)?;
```

Comment:
We'll have to think later about the right behavior here when files can be unlinked (== inodes can disappear).